File: PROTOTYPE/tango.py
from collections import Counter

result=[]
six=[i for i in range(6)]

def display(grid):
    l=len(grid)
    for i in range(l):
        if i%6==0:
            print()
        print(grid[i],end=" ")
    print('\n---------------------------')

def check_row(grid,idx):
    r=get_row(idx)
    hash=Counter([grid[i] for i in r])
    if hash[1]>3 or hash[2]>3:
        return False
    return True


def check_col(grid,idx):
    c=get_col(idx)
    hash=Counter([grid[i] for i in c])
    if hash[1]>3 or hash[2]>3:
        return False
    return True
    

def get_row(idx):
    n=idx//6
    return [n*6+i for i in six]

def get_col(idx):
    n=idx%6
    return [6*i+n for i in six]

def solver(grid,equal,cross,idx,locked):
    global result
    if idx==36:
        display(grid)
        result=grid[:]
        return
    if grid[idx]==0:
        for i in [1,2]:
            grid[idx]=i

            modified_equal=[]
            if idx in equal:
                for j in equal[idx]:
                    if j not in locked:
                        grid[j]=grid[idx]
                        modified_equal.append(j)

            modified_cross=[]
            if idx in cross:
                for j in cross[idx]:
                    if j not in locked:
                        grid[j]=2 if grid[idx]==1 else 1
                        modified_cross.append(j)

            if check_row(grid,idx) and check_col(grid,idx):
                display(grid)
                solver(grid,equal,cross,idx+1,locked)

            for j in modified_equal:
                if j not in locked:
                    grid[j]=0
            for j in modified_cross:
                if j not in locked:
                    grid[j]=0
            grid[idx]=0
    else:
        if check_row(grid,idx) and check_col(grid,idx):
            display(grid)
            solver(grid,equal,cross,idx+1,locked)
        else:
            if idx not in locked:
                grid[idx]=0
            return

        


if __name__=="__main__":
    '''Board setup'''
    tango=[0,0,0,0,0,0,
            0,0,0,0,0,1,
            0,0,0,0,1,0,
            0,0,0,1,0,2,
            0,0,1,0,1,0,
            0,1,0,1,0,2]
    locked=[i for i in range(36) if tango[i]!=0]
    equal={1:[2,7],6:[7,12]}
    cross={}

    #print(tango)
    #display(tango)
    #print([tango[i] for i in get_row(28)])
    #print([tango[i] for i in get_col(28)])
    
    solver(tango,equal,cross,0,locked)
    print("RESULT:")
    display(result)

File: PROTOTYPE/tango.js
let board = [
    0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0,
    0, 0, 2, 0, 0, 0,
    0, 0, 0, 1, 0, 0,
    0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0
];

let equal = new Map([
    [6, [0]],
    [9, [8]],
    [13, [12]],
    [23, [22]],
    [34, [28]],
    [33, [32]]
]);

let cross = new Map([
    [7, [1]],
    [3, [2]],
    [17, [16]],
    [19, [18]],
    [27, [26]],
    [35, [29]]
]);

let num_str = new Map([
    [1, "Sun"],
    [2, "Moon"]
]);

let str_num = new Map([
    ["Sun", 1],
    ["Moon", 2]
]);
const locked=[14,21]
let size = 6;



console.log("EQUAL:",equal);
console.log("CROSS",cross);
console.log("BOARD",board);
console.log("LOCKED",locked);
console.log("NUM TO STRING",num_str);
console.log("STR TO NUM",str_num);
console.log("SIZE",size);

//------------------------------------------------INPUT--------------------
let display=(board)=>{
    console.log(board.map((val, idx) => (idx % size === 0 ? "\n" : "") + val).join(", "));
    console.log("-----------------------");
}

let result=[];

let get_row=((idx)=>{
    let arr=[];
    let r=Math.floor(idx/6);
    for(let i=0;i<size;i++){
        arr.push(board[r*size+i]);
    }
    return arr;
});
let get_col=((idx)=>{
    let arr=[];
    let c=idx%size;
    for(let i=0;i<size;i++){
        arr.push(board[c+i*size]);
    }
    return arr;
})
function check_count(mat){
    let count1=0;
    let count2=0;
    mat.forEach(el =>{
        if(el===1){
            count1++;
        }else if(el==2){
            count2++;
        }
    });
    for(let i=1;i<mat.length-1;i++){
        if(mat[i]!==0 && mat[i]===mat[i-1] && mat[i]===mat[i+1]){
            return false;
        }
    }
    return count1<=3 && count2<=3;
}
function is_Safe(board,idx){
    let row=get_row(idx);
    let col=get_col(idx);
    if(!(check_count(row) && check_count(col))){
        return false;
    }
    if(equal.has(idx)){
        let na=equal.get(idx);
        for(let i=0;i<na.length;i++){
            if(board[na[i]]!==0 && board[na[i]]!==board[idx]) {
                return false;
            }
        }
    }
    if(cross.has(idx)){
        let na=cross.get(idx);
        for(let i=0;i<na.length;i++){
            if(board[na[i]]!==0 && board[na[i]]===board[idx]) {
                return false;
            }
        }
    }
    return true;
}


function backtrack(idx){
    display(board);
    if(idx===size*size){
        result=structuredClone(board);
        console.log("Hey found it!");
        return true
    }
    if(locked.includes(idx)){
        return backtrack(idx+1);
    }
    for(let i=1;i<3;i++){
        board[idx]=i
        if(is_Safe(board,idx)){
            if(backtrack(idx+1))return true;
        }
        board[idx]=0
    }
    return false;
}

//-------------------CALLING-------------------
display(board);
console.log(backtrack(0))
display(result);
//console.log(get_row(12),get_row(11),get_row(14),get_row(21));
//console.log(get_col(12),get_col(11),get_col(14),get_col(21));
